thunder/core/trace_interpreter.py

```diff
@@ -203,3 +204,167 @@ def do_swap(v):
         return new_trace, tree_map(read, trace.output), env
 
     return new_trace, tree_map(read, trace.output)
+
+
+class TraceSubstitutionProcessor:
+    NULL = object()
+
+    def __init__(self, trace, *args, **kwargs):
+        self.env = {}
+        self.trace = trace
+        self.new_trace = from_trace(self.trace)
+        self.have_processed_args = False
+
+    def read(self, x: VariableInterface | Any) -> Any:
+        if isinstance(x, VariableInterface):
+            return self.env[x.name]
+        else:
+            return x
+
+    def write(self, v: VariableInterface | Any, val: Any, allow_duplicates=True) -> None:
+        if not isinstance(v, VariableInterface):
+            return
+        if v.name in self.env:
+            if allow_duplicates:
+                # Duplicates are allowed and not overwritten
+                return
+            raise ValueError(f"Variable {v.name} is being overwritten this is not allowed")
+        self.env[v.name] = val
+
+    def add_to_swap_map(self, old, new):
+        if old is new:
+            return
+        if isinstance(old, ProxyInterface):
+            if isinstance(new, ProxyInterface) and variableify(new) in self.env:
+                # the new isn't new, but something returned the input
+                # this means we need to map the old to the new
+                old, new = new, old
+            elif isinstance(old, TensorProxyInterface):
```

Comment:
Maybe in the future we won't need to have logic specific to VJPDuals, for example, here

Reply:
The irony is that `VJPDuals` neither have VJP-specific logic nor is there any duality involved, so this should probably just be called `AugmentedValue`, I will eventually send a PR to rename it.
As the next thing is to consolidate `interpret_trace_to_trace` into this, let's keep it for now.